Add goals section to README.md (#971)
xi-editor/druid: README.md

[![docs.rs](https://docs.rs/druid/badge.svg)](https://docs.rs/druid/)
[![license](https://img.shields.io/crates/l/druid)](./LICENSE)

Druid is an experimental Rust-native UI toolkit. Its main goal is to offer a

> polished user experience. There are many factors to this goal, including

Reviewer: To the end user. I think using *user* to mean end-user and *developer* to mean developer is a fine distinction to go with.

> performance, a rich palette of interactions (thus a widget library to support

Reviewer: ~thus~ hence ([*Hence usually refers to the future. Thus usually refers to the past.*](https://www.englishgrammar.org/therefore-hence-so-then-thus-etc/))

[![flex.rs example](https://raw.githubusercontent.com/xi-editor/druid/screenshots/images/0.6.0/flex.png)](./druid/examples/flex.rs)
[![custom_widget.rs example](https://raw.githubusercontent.com/xi-editor/druid/screenshots/images/0.6.0/custom_widget.png)](./druid/examples/custom_widget.rs)

## Goals

Druid's primary goal is to provide a smooth and polished user experience with

> pixel perfect design and no behavioral quirks. In order to archive this we

Reviewer 2: ~archive~ achieve

> strife for a variety of things:

Reviewer: ~strife~ strive

> strive for a variety of things:

Reviewer 3: This feels vague. I might say,

`Druid's goal is to make it easy to write and deploy small and efficient desktop applications that work on most common platforms.'

Reviewer: Sure, but then I'd like to see this item added to the goals list:
* Provide a smooth and polished user experience with pixel perfect design and no behavioral quirks.

smooth = no stutter
polished = care has been given to details
pixel perfect = widgets aren't being drawn willy-nilly in roughly the correct location, but instead in just the right pixels
behavioral quirks = application behaving differently under uncommon scenarios, like hot state not working when alt-tabbing out etc

Author: I tried to strike a middle ground on this one:

*Druid's goal is to make it easy to write and deploy high quality desktop
applications with a smooth and polished user experience on all common
platforms.*

as the summary, and

*Facilitate a smooth and polished user experience with no behavioral quirks*

as an additional point.

Reviewer: I think if we already have the smooth & polish in the summary, then the extra point isn't needed.

Author: Well, you are right. I suppose it was a bit to late when I wrote that and my brain wasn't working properly anymore. I've removed the extra point.

- Make it easy to build and package on all supported platforms.

> - Provide abstractions to avoid platform specific quirks.

Reviewer 3: ```suggestion
- Provide a flexible and ergonomic API that abstracts common platform behaviour.
- Make it possible to interact with the platform directly when more control is necessary.
```

Reviewer: Just to note that *abstracts common platform behaviour* wouldn't cover things like the multi-click api in #859 because that's not a common platform behavior. It's an abstraction that builds the behavior.

> - Encourage good application behavior across platforms.

Reviewer 3: ```suggestion
- Respect platform conventions and expectations.
```

> - Automate platform resolution independence.

Reviewer 3: ```suggestion
- Simple and robust handling for display resolution and scaling.
```

(I'm less sure about this wording)

Reviewer: All the other items start with verbs, so this change would kind of ruin that vibe. :thinking:

Author: Maybe like this:

Handle display resolution and scaling reliably with little effort.

> - Enable robust internationalization.

Reviewer 3: ```suggestion
- Make internationalization easy and robust.
```

> - Ensure accessibility to a wide group of people.

Reviewer 3: this one is tricky; we're just a long way away. I'm not sure if I would include it, but if I would I would just say,
'Robust accessibility support'.

Reviewer: This change would also lose the verb from the beginning. If you don't want to mention people and commitment, then perhaps:
*Provide robust accessibility support.*

Author: We already use `provide` for the abstraction point.
Maybe:
*Offer robust accessibility support.*

- Produce small and fast binaries with low memory usage.

> - Have a small dependency tree with a high quality code base and good organization.

Reviewer: I would say having a small dependency tree is an end goal. To put it differently, having a small dependency tree is one of the major reasons I'm personally interested in druid over Electron. I wrote in more detail about it #806.

Also the point about `cargo build` working is even more of a process goal, because that won't have any impact after the build step, while a small dependency tree will matter at runtime too.

> - Have a small dependency tree with a high quality code base and good organization.

Reviewer 3: I guess I see "produce small and fast binaries with low memory usage" as being our shot at electron?

> - Have a small dependency tree with a high quality code base and good organization.

Author: Aside from being a good goal in general, I think it can be seen as such.

> - Focus on powerful, desktop-grade applications.

Reviewer 3: this is a longer-term goal, and I'm a bit hesitant about including it here; this might be a good discussion though, of like what are a good set of intermediate goals?

For instance, right now I think a good goal would be "make it easy to create a simple UI for your rust application".

Reviewer: I think this was included to contrast with mobile.

Author: Yep, it is about support for complex interactions which are needed by more powerful applications (as which I would classify Runebender for example).

> - Provide the basic set of widgets needed by most applications.

Reviewer 3: also a bit vague.
maybe: 

- Provide a flexible set of built-in widgets
- Make it easy to create custom widgets and application logic as needed

Author: Hmm, how about
- Offer a flexible set of layouts and common widgets.
- Ease creation of custom components and application logic as needed.

- Simplify creation of complex, special purpose widgets.

### Non-Goals

> In order to fulfill those goals, we can not support every use case. Luckily

Reviewer: *can not* is [sort of fine](https://www.lexico.com/grammar/cannot-or-can-not), but more common would be *cannot*

the Rust community is working on a variety of different libraries with
different goals, so here are some of druid's non-goals and possible
alternatives that can offer those capabilities:

- Use the the platform-native widgets or mimic them. ([Relm])
- Embed easily into custom render pipelines. ([Conrod])
- Implement a specific architectural style such as Elm. ([Iced], [Relm])
- Support rendering to HTML when compiling for WASM. ([Iced], [Moxie])

> There is also a collection of many more ongoing [Rust-native GUI experiments].

Reviewer 3: Does this link work? I don't see it when I skim.

```suggestion
Druid is just one of many ongoing [Rust-native GUI experiments]. If it doesn't suit your use case, perhaps one of the others will!
```

Reviewer: The link does work. For reference when viewing the changes on GitHub, you can click on the three dots button in the top right and choose "View file" then that will open that as formatted markdown and you can click on the link.

> There also is a collection of many more ongoing [Rust-native GUI experiments].

Reviewer: ~There also is~ There is also
